### PR TITLE
IONDRVFramebuffer::_doControl denoiser

### PR DESCRIPTION
**Pull request by nms42**
When _doControl() complied with #define IONDRVCHECK 1, it produce a lot of debug noise on every csc{Set,Draw}HardwareCursor action in system logs.

### CODE REVIEW

File: WhateverGreen/kern_ngfx.cpp [hunk 1]
```diff
@@ -27,16 +27,39 @@ static const char *pathNVDAStartupWeb[] {
...
 
+static const char *pathIONDRVSupport[] {
+	"/System/Library/Extensions/IONDRVSupport.kext/IONDRVSupport"
+};
+
 static KernelPatcher::KextInfo kextList[] {
 	{ "com.apple.GeForce", pathGeForce, arrsize(pathGeForce), {}, {}, KernelPatcher::KextInfo::Unloaded },
 	{ "com.nvidia.web.GeForceWeb", pathGeForceWeb, arrsize(pathGeForceWeb), {}, {}, KernelPatcher::KextInfo::Unloaded },
-	{ "com.nvidia.NVDAStartupWeb", pathNVDAStartupWeb, arrsize(pathNVDAStartupWeb), {}, {}, KernelPatcher::KextInfo::Unloaded }
+	{ "com.nvidia.NVDAStartupWeb", pathNVDAStartupWeb, arrsize(pathNVDAStartupWeb), {}, {}, KernelPatcher::KextInfo::Unloaded },
```
Comment (vit9696) [inline WhateverGreen/kern_ngfx.cpp:37]: Why is that needed?

**Comment by nms42** [comment]
On Thu, Feb 25, 2021 at 10:01 PM Vitaly Cheptsov <notifications@github.com>
wrote:

> *@vit9696* commented on this pull request.
> ------------------------------
>
> In WhateverGreen/kern_ngfx.cpp
> <https://github.com/acidanthera/WhateverGreen/pull/83#discussion_r583089394>
> :
>
> >  static KernelPatcher::KextInfo kextList[] {
>  	{ "com.apple.GeForce", pathGeForce, arrsize(pathGeForce), {}, {}, KernelPatcher::KextInfo::Unloaded },
>  	{ "com.nvidia.web.GeForceWeb", pathGeForceWeb, arrsize(pathGeForceWeb), {}, {}, KernelPatcher::KextInfo::Unloaded },
> -	{ "com.nvidia.NVDAStartupWeb", pathNVDAStartupWeb, arrsize(pathNVDAStartupWeb), {}, {}, KernelPatcher::KextInfo::Unloaded }
> +	{ "com.nvidia.NVDAStartupWeb", pathNVDAStartupWeb, arrsize(pathNVDAStartupWeb), {}, {}, KernelPatcher::KextInfo::Unloaded },
>
> Why is that needed?
>

Due to comma in list

**Comment by vit9696** [comment]
Yes, realised it. I do not like the patch the way it is done, and suggest two alternatives to be prone to future updates:

- Reimplement the function ourselves, it looks really tiny, and I do not believe the offsets change.
- Lookup the function and the print function, and then search for the relevant call by calculating the offset.

What do you think?

**Comment by nms42** [comment]
> * Reimplement the function ourselves, it looks really tiny, and I do not believe the offsets change.
> * Lookup the function and the print function, and then search for the relevant call by calculating the offset.

What would be gained? Current version is a kind of #define IONDRVCHECK 0. Reimplementation brings chasing possible mods of original sources by Apple.

May be KernelPatcher::patchFunction() in Lilu? Masked patching also?

**Comment by vit9696** [comment]
Current patch would break and we do not want extra maintenance burden. It also is designed to work on Catalina only, while it clearly applies to older versions. Apple is unlikely to mod much as they have new graphics stack these days.

**Comment by nms42** [comment]
> Current patch would break and we do not want extra maintenance burden. It also is designed to work on Catalina only, while it clearly applies to older versions. Apple is unlikely to mod much as they have new graphics stack these days.

This patch is for BigSur (there is >Catalina check) only because in Catalina and older versions _doControl() compiled with #define IONDRVCHECK 0.

**Comment by vit9696** [comment]
Hrm. Could you patch this method with ud2 (0F 0B) and show a backtrace with symbols?

**Comment by nms42** [comment]
> Hrm. Could you patch this method with ud2 (0F 0B) and show a backtrace with symbols?

What you looking for?

**Comment by vit9696** [comment]
For the reason this method is frequently called in the first place in this particular case.

**Comment by nms42** [comment]
> For the reason this method is frequently called in the first place in this particular case.

Callers are in the same source file mentioned in patch. All of noise come from calls on every mouse cursor change -- codes 22&23 == csc{Set|Draw}HardwareCursor. As I understand this action can be done without gating.

**Comment by vit9696** [comment]
It is exactly the opposite. The message you mentioned can only be printed when the function is called without gating, i.e. it is a mistake in the driver to call this function directly. Printing a backtrace would help us locate the place, and perhaps find a better fix.

For that you can route the function through a proxy and write something like that:

 ```C++
    IOWorkLoop * wl;
    if (!ml_at_interrupt_context() && (wl = fb->getControllerWorkLoop()) && !wl->inGate())
      PANIC("weg", "called _doControl outside of workloop");
```

**Comment by nms42** [comment]
> It is exactly the opposite. The message you mentioned can only be printed when the function is called without gating, i.e. it is a mistake in the driver to call this function directly.

It is not a mistake because above mentioned cursor operations can be done without gating. cscSetHardwareCursor & cscDrawHardwareCursor are special cases unaccounted in this particular check.

**Comment by nms42** [comment]
HardwareCursor control is frequently performed thus made gateless on purpose, I understand.

**Comment by vit9696** [comment]
I am not sure about it, but not impossible. I updated the code, please test it. If it works, we will merge.

Please also run this code and provide a kernel panic:

```C++
IOReturn NGFX::wrapNdrvDoControl(IONDRVFramebuffer *fb, UInt32 code, void *params) {
	// Suppress debug(?) noise from IONDRVFramebuffer::_doControl( IONDRVFramebuffer * fb, UInt32 code, void * params )
	// when function was compiled with #define IONDRVCHECK 1.
	// Ref: https://opensource.apple.com/source/IOGraphics/IOGraphics-585/IONDRVSupport/IONDRVFramebuffer.cpp.auto.html

	if (code == cscSetHardwareCursor || code == cscDrawHardwareCursor) {
		if (NDRVFramebufferViewer::getState(fb) == 0)
			return kIOReturnNotOpen;

		IOWorkLoop * wl;
		if (!ml_at_interrupt_context() && (wl = fb->getControllerWorkLoop()) && !wl->inGate())
			PANIC("ngfx", "IONDRVFramebuffer::_doControl(%d) not gated\n", code);

		IONDRVControlParameters pb;
		pb.code = code;
		pb.params = params;
		return fb->doDriverIO(/*ID*/ 1, &pb, kIONDRVControlCommand, kIONDRVImmediateIOCommandKind);
	}

	return FunctionCast(wrapNdrvDoControl, callbackNGFX->orgNdrvDoControl)(fb, code, params);
}

**Comment by nms42** [comment]
> I am not sure about it, but not impossible. I updated the code, please test it. If it works, we will merge.
> Please also run this code and provide a kernel panic:

No noise/messages with this version. Publish binary on applelife.ru for more testers?

**Comment by vit9696** [comment]
Could do, but please provide a panic log as well.

**Comment by nms42** [comment]
> Could do, but please provide a panic log as well.

What kind of panic? Elaborate, please

**Comment by vit9696** [comment]
Well, I explained in https://github.com/acidanthera/WhateverGreen/pull/83#issuecomment-787051784?

**Comment by nms42** [comment]
Is this sufficient? Due to timing one can get panic with any of calls.

```
nms@Steely Desktop % fgrep _doControl IONDRVFramebuffer.cpp
            err = fb->_doControl( fb, cscProbeConnection, 0 );
					next->_doControl( next, cscProbeConnection, 0 );
    err = _doControl( this, cscSetHardwareCursor, &setCursor );
        err = _doControl( this, cscDrawHardwareCursor, &drawCursor );
        err = _doControl( fb, (UInt32)(uintptr_t) code, params );
                    err = _doControl( this, cscDoCommunication, &comm );
        _doControl( this, cscGrayPage, &pageInfo);
                _doControl( this, cscSetClutBehavior, &clutSetting));
            err = _doControl( this, cscSetDetailedTiming, &look );
        err = _doControl( this, cscSetDetailedTiming, &newTiming );
            err = _doControl( this, cscSetScaler, &scaler );
    err = _doControl( this, cscSetGray, &grayRec );
        _doControl( this, cscSetMode, &pageInfo);
        _doControl( this, cscGrayPage, &pageInfo);
    err = _doControl( this, cscSwitchMode, &switchInfo);
    _doControl( this, cscSetMode, &pageInfo);
            _doControl( this, cscSetGamma, &gammaRec );
            _doControl( this, cscGrayPage, &pageInfo);
        _doControl( this, cscGrayPage, &pageInfo);
    err = _doControl( this, cscSavePreferredConfiguration, &switchInfo);
        _doControl( this, cscSetGray, &grayRec);
        _doControl( this, cscSetClutBehavior, &clutSetting);
    err = _doControl( this, cscSetEntries, &setEntryRec);
            _doControl( this, cscSetClutBehavior, &clutSetting);
    err = _doControl( this, cscSetGamma, &gammaRec );
        err = _doControl( this, cscSetGamma, &gammaRec );
        err = _doControl( this, cscSetMirror, &mirror );
        err = _doControl(this, cscDoCommunication, &comm);
    if ((noErr == _doControl(this, cscDoCommunication, &i2CRecord))
        _doControl(this, cscDoCommunication, &i2CRecord);
            err = _doControl( this, cscSetFeatureConfiguration, &configRec );
            _doControl( this, cscSetSync, &syncInfo);
                ret = _doControl( this, cscProbeConnection, 0 );
        _doControl( this, cscSetSync, &syncInfo);
        err = _doControl(this, cscSetPowerState, &sleepInfo);
            err = _doControl( this, cscSleepWake, &sleepInfo);
            err = _doControl( this, cscSetPowerState, &sleepInfo);
nms@Steely Desktop %
```

**Comment by vit9696** [comment]
Well, obviously not, as the caller is within closed-source NVIDIA driver, and I would like to see at least one backtrace containing NVIDIA symbols?

**Comment by nms42** [comment]
> Well, obviously not, as the caller is within closed-source NVIDIA driver, and I would like to see at least one backtrace containing NVIDIA symbols?

See attachment

[20210228093006.log.zip](https://github.com/acidanthera/WhateverGreen/files/6056358/20210228093006.log.zip)

**Comment by vit9696** [comment]
Hm, thanks for the log, but could you please use the code snippet from https://github.com/acidanthera/WhateverGreen/pull/83#issuecomment-787051784 instead of a ud2 patch as I asked you above? Maybe you did not notice, but I put a PANIC call there right when the call happens outside of the call gate and has the "spam" commands. The panic log you sent me seems to contain a different command.

**Comment by nms42** [comment]
> Hm, thanks for the log, but could you please use the code snippet from [#83 (comment)](https://github.com/acidanthera/WhateverGreen/pull/83#issuecomment-787051784) instead of a ud2 patch as I asked you above? Maybe you did not notice, but I put a PANIC call there right when the call happens outside of the call gate and has the "spam" commands. The panic log you sent me seems to contain a different command.

Sorry, missed the point. No panic (on short time test). Just stream of this

```
Backtrace 0xffffff7fa0aa9b34 0xffffff7fa0a767bb 0xffffff7fa0a77464 0xffffff8000a4628e 0xffffff7fa0a869e2 0xffffff8000a5045b 0xffffff80003a8e51
      Kernel Extensions in backtrace:
         com.apple.iokit.IOGraphicsFamily(585.0)[EF8876FC-C6D6-3D53-9915-762EF4F12870]@0xffffff7fa0a62000->0xffffff7fa0a90fff
            dependency: com.apple.iokit.IOPCIFamily(2.9)[BF2C5E86-1E8F-3FD4-9874-7738178FA73B]@0xffffff8002c1f000->0xffffff8002c46fff
         com.apple.iokit.IONDRVSupport(585.0)[40BEBEA2-BBB5-3B09-B6AE-B4B372775FBD]@0xffffff7fa0aa7000->0xffffff7fa0aaffff
            dependency: com.apple.iokit.IOGraphicsFamily(585)[EF8876FC-C6D6-3D53-9915-762EF4F12870]@0xffffff7fa0a62000->0xffffff7fa0a90fff
            dependency: com.apple.iokit.IOPCIFamily(2.9)[BF2C5E86-1E8F-3FD4-9874-7738178FA73B]@0xffffff8002c1f000->0xffffff8002c46fff
GFX0-0x22c-B::_doControl(23) not gated

Backtrace 0xffffff7fa0aa9b34 0xffffff7fa0a767bb 0xffffff7fa0a77464 0xffffff8000a4628e 0xffffff7fa0a869e2 0xffffff8000a5045b 0xffffff80003a8e51
      Kernel Extensions in backtrace:
         com.apple.iokit.IOGraphicsFamily(585.0)[EF8876FC-C6D6-3D53-9915-762EF4F12870]@0xffffff7fa0a62000->0xffffff7fa0a90fff
            dependency: com.apple.iokit.IOPCIFamily(2.9)[BF2C5E86-1E8F-3FD4-9874-7738178FA73B]@0xffffff8002c1f000->0xffffff8002c46fff
         com.apple.iokit.IONDRVSupport(585.0)[40BEBEA2-BBB5-3B09-B6AE-B4B372775FBD]@0xffffff7fa0aa7000->0xffffff7fa0aaffff
            dependency: com.apple.iokit.IOGraphicsFamily(585)[EF8876FC-C6D6-3D53-9915-762EF4F12870]@0xffffff7fa0a62000->0xffffff7fa0a90fff
            dependency: com.apple.iokit.IOPCIFamily(2.9)[BF2C5E86-1E8F-3FD4-9874-7738178FA73B]@0xffffff8002c1f000->0xffffff8002c46fff
GFX0-0x22c-B::_doControl(23) not gated

Backtrace 0xffffff7fa0aa9b34 0xffffff7fa0a767bb 0xffffff7fa0a77464 0xffffff8000a4628e 0xffffff7fa0a869e2 0xffffff8000a5045b 0xffffff80003a8e51
      Kernel Extensions in backtrace:
         com.apple.iokit.IOGraphicsFamily(585.0)[EF8876FC-C6D6-3D53-9915-762EF4F12870]@0xffffff7fa0a62000->0xffffff7fa0a90fff
            dependency: com.apple.iokit.IOPCIFamily(2.9)[BF2C5E86-1E8F-3FD4-9874-7738178FA73B]@0xffffff8002c1f000->0xffffff8002c46fff
         com.apple.iokit.IONDRVSupport(585.0)[40BEBEA2-BBB5-3B09-B6AE-B4B372775FBD]@0xffffff7fa0aa7000->0xffffff7fa0aaffff
            dependency: com.apple.iokit.IOGraphicsFamily(585)[EF8876FC-C6D6-3D53-9915-762EF4F12870]@0xffffff7fa0a62000->0xffffff7fa0a90fff
            dependency: com.apple.iokit.IOPCIFamily(2.9)[BF2C5E86-1E8F-3FD4-9874-7738178FA73B]@0xffffff8002c1f000->0xffffff8002c46fff
GFX0-0x22c-B::_doControl(23) not gated

Backtrace 0xffffff7fa0aa9b34 0xffffff7fa0a767bb 0xffffff7fa0a77464 0xffffff8000a4628e 0xffffff7fa0a869e2 0xffffff8000a5045b 0xffffff80003a8e51
      Kernel Extensions in backtrace:
         com.apple.iokit.IOGraphicsFamily(585.0)[EF8876FC-C6D6-3D53-9915-762EF4F12870]@0xffffff7fa0a62000->0xffffff7fa0a90fff
            dependency: com.apple.iokit.IOPCIFamily(2.9)[BF2C5E86-1E8F-3FD4-9874-7738178FA73B]@0xffffff8002c1f000->0xffffff8002c46fff
         com.apple.iokit.IONDRVSupport(585.0)[40BEBEA2-BBB5-3B09-B6AE-B4B372775FBD]@0xffffff7fa0aa7000->0xffffff7fa0aaffff
            dependency: com.apple.iokit.IOGraphicsFamily(585)[EF8876FC-C6D6-3D53-9915-762EF4F12870]@0xffffff7fa0a62000->0xffffff7fa0a90fff
            dependency: com.apple.iokit.IOPCIFamily(2.9)[BF2C5E86-1E8F-3FD4-9874-7738178FA73B]@0xffffff8002c1f000->0xffffff8002c46fff
GFX0-0x22c-B::_doControl(23) not gated
```

**Comment by vit9696** [comment]
Hmm, the original has these swapped:

```
		if (NDRVFramebufferViewer::getState(fb) == 0)
			return kIOReturnNotOpen;

		IOWorkLoop * wl;
		if (!ml_at_interrupt_context() && (wl = fb->getControllerWorkLoop()) && !wl->inGate())
			PANIC("ngfx", "IONDRVFramebuffer::_doControl(%d) not gated\n", code);
```

i.e.

```
		IOWorkLoop * wl;
		if (!ml_at_interrupt_context() && (wl = fb->getControllerWorkLoop()) && !wl->inGate())
			PANIC("ngfx", "IONDRVFramebuffer::_doControl(%d) not gated\n", code);

		if (NDRVFramebufferViewer::getState(fb) == 0)
			return kIOReturnNotOpen;
```

Maybe try it?

**Comment by nms42** [comment]
Strange, looks like reroute does not working.

See new branch doCntrol-panicer and relevant stuff in attachment
[dC-panicer.zip](https://github.com/acidanthera/WhateverGreen/files/6056741/dC-panicer.zip)

**Comment by vit9696** [comment]
Indeed this is quite strange. Can you test with this build?

[WhateverGreen-1.4.8-DEBUG.zip](https://github.com/acidanthera/WhateverGreen/files/6056892/WhateverGreen-1.4.8-DEBUG.zip)

**Comment by nms42** [comment]
The same stuff )-; Just small mouse move

[20210228182817.log.zip](https://github.com/acidanthera/WhateverGreen/files/6057023/20210228182817.log.zip)

**Comment by vit9696** [comment]
Well, but the version in the PR works for you, right?

**Comment by nms42** [comment]
> Well, but the version in the PR works for you, right?

Yes. Version with rerouting but without PANIC works. Complier optimized out all the
```
if (code == ... || code == ...) {
}
```
??? Can it be?

**Comment by vit9696** [comment]
Do not think so. I do not see WhateverGreen at all in your log. Maybe your Lilu version is outdated and it simply does not load?

**Comment by vit9696** [comment]
I merged the PR to fit it in the release, but I still want the panic.

**Comment by nms42** [comment]
> I merged the PR to fit it in the release, but I still want the panic.

Recompiled Lilu & WEG.

Tryed, still no panic. Just backtrace records in log. See 11.log in attachment.

Tryed with yours WEG binaries. Just backtrace records in log. See 22.log in attachment.

[nlogs.zip](https://github.com/acidanthera/WhateverGreen/files/6061525/nlogs.zip)

**Comment by vit9696** [comment]
Well, I do not see WEG loaded in this log. Does it show in kextstat?

**Comment by nms42** [comment]
> Well, I do not see WEG loaded in this log. Does it show in kextstat?

I discovered that too. Preparing debug logs.

/Stay tuned!

**Comment by nms42** [comment]
Attached log and config.plist used.
[noweg.zip](https://github.com/acidanthera/WhateverGreen/files/6062396/noweg.zip)

**Comment by vit9696** [comment]
14:836 00:066 OC: Prelinked injection WhateverGreen.kext () - Invalid Parameter. Are you sure USB flash is not corrupted? @Andrey1970AppleLife https://github.com/acidanthera/WhateverGreen/pull/83#issuecomment-787457363 does this inject for you?

**Comment by nms42** [comment]
Yep, over the last days two sticks were used. Just rechecked with diskutil the latest one. Still no injection.

**Comment by nms42** [comment]
> 14:836 00:066 OC: Prelinked injection WhateverGreen.kext () - Invalid Parameter. Are you sure USB flash is not corrupted? @Andrey1970AppleLife [#83 (comment)](https://github.com/acidanthera/WhateverGreen/pull/83#issuecomment-787457363) does this inject for you?

Is there a way to increase debug verbosity for more details. My config.plist in noweg.zip.

**Comment by vit9696** [comment]
this is really the maximum, please attach your kernel and EFI folder with kexts for testing in userspace.

**Comment by nms42** [comment]
What is the 'kernel'? Where to find it? EFI attached.

[EFI.zip](https://github.com/acidanthera/WhateverGreen/files/6063146/EFI.zip)

**Comment by vit9696** [comment]
`/System/Library/KernelCollections/BootKernelExtensions.kc`

**Comment by nms42** [comment]
Here is kernel

[BootKernelExtensions.kc.zip](https://github.com/acidanthera/WhateverGreen/files/6063205/BootKernelExtensions.kc.zip)

**Comment by vit9696** [comment]
Seems to be caused by missing IOGraphicsFamily dependency. Try solving getWorkloop symbol manually from com.apple.iokit.IOGraphicsFamily through patcher:

```
OCAK: Symbol __ZNK13IOFramebuffer21getControllerWorkLoopEv was resolved for kext as.vit9696.WhateverGreen to 0

**Comment by nms42** [comment]
That is beyond my comprehension )-; Can you make such binary for me, please?

**Comment by vit9696** [comment]
Recorded this in my backlog.
